docs/notebooks/documents/service_manager.py

```diff
@@ -0,0 +1,73 @@
+# ---
+# jupyter:
+#   jupytext:
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.16.1
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Service Management for PDO Contracts (WIP)
+#
+# This notebook helps to manage the services database.
+
+# %%
+import pdo.contracts.jupyter as pc_jupyter
+import IPython.display as ip_display
+import ipywidgets
+
+from pdo.contracts.services import ServiceUploadWidget, ServiceSelectionWidget, ServiceListWidget
+from pdo.contracts.services import service_labels, get_by_url, update_service
+
+pc_jupyter.load_ipython_extension(get_ipython())
+
+try :
+    state
+except NameError :
+    (state, bindings) = pc_jupyter.initialize_environment('unknown')
+
+# %% [markdown]
+# ## Import Services
+#
+# Typically, a site will generate a file with information about the service endpoints. The file is
+# usually called `site.toml`. Upload a `site.toml` file to import services.
+# %%
+ip_display.display(ServiceUploadWidget(state, bindings))
+
+# %% [markdown]
+# ## Services List
+#
+# Display a list of the currently registered services. Services can be imported
+# from a site configuration file below.
+# %%
+children = map(lambda stype : ServiceListWidget(state, bindings, stype), service_labels.keys())
+service_list_widget = ipywidgets.Tab(children=list(children), titles=list(service_labels.values()))
+ip_display.display(service_list_widget)
+
+# %% [markdown]
+# ## Verify Services
+#
+# Verification checks that the service exists and that its identity has not
+# changed since it was added to the service database. While correct operation
+# does not require service verification, it is often useful to check on the
+# status of services.
+# %%
+def verify_service(service_type, urls, output) :
```

Comment:
"verification" term is a bit overloaded/confusing. I suppose within the esrvicedb, we verify each service against ledger, and record "ledger verified time in the db". whereas looks like in here "verify" means a "healthcheck" of the service itself that the client gets (via the /info url). how are these two verifications related?

Reply:
in this case, not really. the verify function does exactly that... it makes sure that the service exists and that the identity has not changed. this is old code.

Reply:
ok thanks, so if I understand correctly, verify here does not verify against the ledger, just a healthcheck of the service. 
thinking aloud, for a contract user, sepcifically one who just wants to participate in an already deployed  PDO-based application (such as new token owner), what would be useful is for the user to start with the application context file, and invoke a verify API on the context file, and behind the scenes, the validity of the services specified by the whole context file gets verified (including ledger checks for attestation/registry info, service heartbeats such as what you are doing here etc.). We should not be expecting the user (I mean the new token owner) to maintain (or at least understand) any other config file other this user needs to maintain. PDO specific configs such as servicedbs, groupdbs must be constructed transparently based on the info found in context.